### PR TITLE
presto support

### PR DESCRIPTION
**Pull request by zmz0305**

### CODE REVIEW

File: fili-sql/src/main/java/com/yahoo/bard/webservice/sql/DefaultSqlBackedClient.java [hunk 1]
```diff
@@ -127,6 +127,7 @@ private JsonNode executeAndProcessQuery(DruidAggregationQuery<?> druidQuery) {
                 druidQueryToSqlConverter.getTimeConverter()
         );
 
+        LOG.info("getConnection in executeAndProcessQuery");
```
Comment (zmz0305) [inline fili-sql/src/main/java/com/yahoo/bard/webservice/sql/DefaultSqlBackedClient.java]: remove

File: fili-sql/pom.xml [hunk 2]
```diff
@@ -29,6 +29,8 @@
 
         <version.h2>1.4.195</version.h2>
         <version.calcite>1.20.0</version.calcite>
+        <presto.version>0.204.2.4.1909232158</presto.version>
+        <!--<presto.version>0.0.1</presto.version>-->
```
Comment (zmz0305) [inline fili-sql/pom.xml]: remove

File: fili-sql/pom.xml [hunk 3]
```diff
@@ -50,6 +52,22 @@
             <version>${version.h2}</version>
             <scope>test</scope>
         </dependency>
+        <!--<dependency>-->
```
Comment (zmz0305) [inline fili-sql/pom.xml]: remove

File: fili-sql/src/main/java/com/yahoo/bard/webservice/sql/helper/CalciteHelper.java [hunk 4]
```diff
@@ -36,6 +49,7 @@
     public CalciteHelper(DataSource dataSource) throws SQLException {
         this.dataSource = dataSource;
 
+        LOG.info("getConnection in CalciteHelper constructor");
```
Comment (zmz0305) [inline fili-sql/src/main/java/com/yahoo/bard/webservice/sql/helper/CalciteHelper.java]: remove

File: fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/PrestoResultSetProcessor.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    ) throws IOException {
+        int lastTimeIndex = sqlTimeConverter.timeGrainToDatePartFunctions(druidQuery.getGranularity()).size();
+        int columnCount = columnToColumnName.size();
+
+        for (int i = 0; i < columnCount; i++) {
+            if (isTimeColumn(lastTimeIndex, i)) {
+                continue;
+            }
+            String columnName = columnToColumnName.get(i);
+            Boolean containsUnderscore = columnName.contains("_");
```
Comment (zmz0305) [inline fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/PrestoResultSetProcessor.java]: simplify this

File: fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/PrestoResultSetProcessor.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+            String tmpcolumnName = "";
+            if (containsUnderscore) {
+                tmpcolumnName = CaseFormat.LOWER_UNDERSCORE.to(CaseFormat.LOWER_CAMEL, columnName);
+            } else {
+                tmpcolumnName = columnName;
+            }
+            if (resultTypeMapper.containsKey(columnName)) {
+                String elementAti = row[i] == null ? "0" : row[i];
```
Comment (zmz0305) [inline fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/PrestoResultSetProcessor.java]: Double check the "0" default value.

File: fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/DruidQueryToPrestoConverter.java [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+    /**
+     * Converts the druid query to a {@link RelNode}.
+     *
+     * @param druidQuery  The query to convert to sql.
+     * @param apiToFieldMapper  The mapping between api and physical names for the query.
+     * @param sqlTable  The sql table being queried against.
+     *
+     * @return the sql equivalent of the query.
+     */
+    private RelNode convertDruidQueryToRelNode(
```
Comment (zmz0305) [inline fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/DruidQueryToPrestoConverter.java]: add @Override

File: fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/DruidQueryToPrestoConverter.java [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+     * Builds a converter between druid and sql aggregations.
+     *
+     * @return a new druid to sql aggregation converter.
+     */
+    protected BiFunction<Aggregation, ApiToFieldMapper, Optional<SqlAggregation>> buildDruidSqlTypeConverter() {
+        return new DruidSqlAggregationConverter();
+    }
+
+    /**
+     * Converts the druid query to a {@link RelNode}.
```
Comment (zmz0305) [inline fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/DruidQueryToPrestoConverter.java]: Add comment about what's different from base class

File: fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/DruidQueryToPrestoConverter.java [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+                        NO_OFFSET,
+                        getLimit(druidQuery),
+                        getSort(builder, druidQuery, apiToFieldMapper, sqlTable.getTimestampColumn())
+                )
+                .build();
+    }
+
+
+    /**
+     * Finds the sorting for a druid query.
```
Comment (zmz0305) [inline fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/DruidQueryToPrestoConverter.java]: add comment about what's different from the parent class

File: fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/DruidQueryToPrestoConverter.java [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+        if (druidQuery.getQueryType().equals(GROUP_BY)) {
+            GroupByQuery groupByQuery = (GroupByQuery) druidQuery;
+            LimitSpec limitSpec = groupByQuery.getLimitSpec();
+            if (limitSpec != null) {
+                limitSpec.getColumns()
+                        .stream()
+                        .map(orderByColumn -> {
+                            String orderByField = apiToFieldMapper.unApply(orderByColumn.getDimension());
+                            limitSpecColumns.add(orderByField);
+                            RexNode sort = builder.literal(orderByField); //presto fix
```
Comment (zmz0305) [inline fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/DruidQueryToPrestoConverter.java]: Find out the reason why we are using literal

File: fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/DruidQueryToPrestoConverter.java [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+        sorts.addAll(builder.fields().subList(druidQuery.getDimensions().size(), groupBys));
+
+        // add limit spec group by
+        sorts.addAll(limitSpecSorts);
+
+        // add remaining group by
+        List<RexNode> unorderedDimensions = druidQuery.getDimensions().stream()
+                .map(Dimension::getApiName)
+                .map(apiToFieldMapper)
+                .filter(columnName -> !limitSpecColumns.contains(columnName))
```
Comment (zmz0305) [inline fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/DruidQueryToPrestoConverter.java]: Confirm the filter logic here

File: fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/DruidQueryToPrestoConverter.java [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+                .map(optionalSqlAggregation -> optionalSqlAggregation.orElseThrow(() -> {
+                    String msg = "Couldn't build sql aggregation with " + optionalSqlAggregation;
+                    LOG.debug(msg);
+                    return new RuntimeException(msg);
+                }))
+                .map(sqlAggregation -> builder.aggregateCall(
+                        sqlAggregation.getSqlAggFunction(),
+                        false,
+                        null,
+                        CaseFormat.LOWER_UNDERSCORE.to(
```
Comment (zmz0305) [inline fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/DruidQueryToPrestoConverter.java]: Call out the difference in comment

File: fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/DruidQueryToPrestoConverter.java [hunk 13]
```diff
@@ -0,0 +1,318 @@
...
+     * @param apiToFieldMapper api column to logic column name mapping
+     * @return a list of RexNode representing the post aggregation
+     */
+    private List<RexNode> getPostAggregations(
+            RelBuilder builder,
+            DruidAggregationQuery<?> druidQuery,
+            ApiToFieldMapper apiToFieldMapper
+    ) {
+        DruidPostAggregationToPresto druidPostAggregationToSql = new DruidPostAggregationToPresto();
+        List<RexNode> fields = new ArrayList<>();
```
Comment (zmz0305) [inline fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/DruidQueryToPrestoConverter.java]: change fields to postAggregation

File: fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/PrestoDataSource.java [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+
+    /**
+     * Constructor.
+     * Follow https://docs.google.com/presentation/d/17QfH9uF6JzTkbA8SSVhS-xwqKfpEEtgDxqYqMlK3B4s/edit#slide=id.p
+     *
+     * TODO: change user to headless
+     */
+    public PrestoDataSource() {
+        properties = new Properties();
+//        properties.setProperty("user", "mzhao50");
```
Comment (zmz0305) [inline fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/PrestoDataSource.java]: make into system config

File: fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/PrestoResultSetProcessor.java (hunk 5 above)
Comment (zmz0305) [reply fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/PrestoResultSetProcessor.java]: resolved

File: fili-sql/pom.xml (hunk 2 above)
Comment (zmz0305) [reply fili-sql/pom.xml]: resolved

File: fili-sql/pom.xml (hunk 3 above)
Comment (zmz0305) [reply fili-sql/pom.xml]: resolved

File: fili-sql/src/main/java/com/yahoo/bard/webservice/sql/DefaultSqlBackedClient.java (hunk 1 above)
Comment (zmz0305) [reply fili-sql/src/main/java/com/yahoo/bard/webservice/sql/DefaultSqlBackedClient.java]: resolved

File: fili-sql/src/main/java/com/yahoo/bard/webservice/sql/helper/CalciteHelper.java (hunk 4 above)
Comment (zmz0305) [reply fili-sql/src/main/java/com/yahoo/bard/webservice/sql/helper/CalciteHelper.java]: resolved

File: fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/DruidQueryToPrestoConverter.java (hunk 7 above)
Comment (zmz0305) [inline fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/DruidQueryToPrestoConverter.java]: **This function is not actually called due to improper class inheritance**

File: fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/DruidQueryToPrestoConverter.java (hunk 8 above)
Comment (zmz0305) [reply fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/DruidQueryToPrestoConverter.java]: resolved

File: fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/DruidQueryToPrestoConverter.java (hunk 7 above)
Comment (zmz0305) [reply fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/DruidQueryToPrestoConverter.java]: Resolved

File: fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/DruidQueryToPrestoConverter.java (hunk 9 above)
Comment (zmz0305) [reply fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/DruidQueryToPrestoConverter.java]: Resolved

File: fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/PrestoResultSetProcessor.java (hunk 6 above)
Comment (zmz0305) [reply fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/PrestoResultSetProcessor.java]: Addressed with ```               if (row[i] == null) {
                    jsonWriter.writeNullField(tmpcolumnName);
                }
                Number result = resultTypeMapper
                        .get(columnName)
                        .apply(row[i]);
                writeNumberField(jsonWriter, tmpcolumnName, result);```

File: fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/DruidQueryToPrestoConverter.java (hunk 11 above)
Comment (zmz0305) [reply fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/DruidQueryToPrestoConverter.java]: limitSpecColumns is returned in `getLimit()`, which is called before `getSort()`, so the columns need to be excluded from getSort()

File: fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/DruidQueryToPrestoConverter.java (hunk 12 above)
Comment (zmz0305) [reply fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/DruidQueryToPrestoConverter.java]: Resolved.

File: fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/DruidQueryToPrestoConverter.java (hunk 13 above)
Comment (zmz0305) [reply fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/DruidQueryToPrestoConverter.java]: Resolved.

File: fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/PrestoDataSource.java (hunk 14 above)
Comment (zmz0305) [reply fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/PrestoDataSource.java]: Resolved

File: fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/DruidQueryToPrestoConverter.java (hunk 10 above)
Comment (zmz0305) [reply fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/DruidQueryToPrestoConverter.java]: Looks like when the code is written, the Calcite lib is an old version. Tested the current version both literal and filed works the same.

File: fili-sql/pom.xml (hunk 2 above)
Comment (colemanProjects) [reply fili-sql/pom.xml]: could we remove this?

File: fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/DruidQueryToPrestoConverter.java (hunk 8 above)
Comment (colemanProjects) [reply fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/DruidQueryToPrestoConverter.java]: `method` typo

File: fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/DruidQueryToPrestoConverter.java (hunk 10 above)
Comment (colemanProjects) [reply fili-sql/src/main/java/com/yahoo/bard/webservice/sql/presto/DruidQueryToPrestoConverter.java]: 👍

**Pull request merged**